### DISCUSSION TITLE
Is it possible to customize error response messages?

### DISCUSSION THREAD

**Discussion by Jay-fly** [fastapi-users/fastapi-users#1470] [Q&A] [answered]
Hello,

I was wondering if there’s a way to customize the error messages returned by the API. For example, in the `/login` route, when user authentication fails, the response currently looks like this:

```json
{  
  "detail": "LOGIN_BAD_CREDENTIALS"  
}  
```
I would like to modify it to something like this:
```json
{  
  "code": "LOGIN_BAD_CREDENTIALS",  
  "reason": "Invalid username or password."  
}  
```

If the only way to achieve this is by overriding the current implementation, do you have recommendations for how to properly rewrite responses for cases like 401 errors? For instance:

Current response for 401:
```json
{  
  "detail": "Unauthorized"  
}  
```
Desired custom response:
```json
{  
  "code": "UNAUTHORIZED",  
  "reason": "Unauthorized"  
}  
```
Thank you for your help!

**Comment by hgalytoby** [comment, marked as answer]
This is a similar situation with inconsistent responses.

https://github.com/fastapi-users/fastapi-users/issues/1318

If you want to standardize all the responses to

```
{  
  “code": ”...” ,  
  “reason": ”...”  
}  
```

I think it's feasible to make a middleware to handle this.

**Reply by Jay-fly** [reply]
Thank you for your suggestion! I implemented the middleware as you recommended, and it works well.

```python
@app.exception_handler(HTTPException)
    async def http_exception_handler(request: Request, exc: HTTPException) -> JSONResponse:

        detail = exc.detail
        error_obj = None

        if isinstance(detail, str):

            if detail == "Forbidden":
                error_obj = ErrorCode.Permission.PERMISSION_DENIED

            elif detail == "Unauthorized":
                error_obj = ErrorCode.Auth.UNAUTHORIZED

            elif detail == "Not Found":
                error_obj = ErrorCode.Entity.ENTITY_NOT_EXIST

            else:
                error_obj = fastapi_users_error_mapping.get(detail, None)

        elif isinstance(detail, dict):
            code = detail.get("code", None)

            error_obj = fastapi_users_error_mapping.get(code, None)

        if error_obj is None:
            # TODO Log here

            error_obj = ErrorCode.Auth.INVALID_TOKEN

        return JSONResponse(
            status_code=exc.status_code,
            content={"code": error_obj.code, "reason": error_obj.reason},
        )
```

**Reply by hgalytoby** [reply]
That's great!

Keep in mind that the downside of this approach is that the “swagger response” information will be incorrect!
